### ISSUE TITLE
Support normalization like in PyTorch

### ISSUE THREAD

**Issue by sedletsky-f5** [knights-analytics/hugot#20]
When we run `NewFeatureExtractionPipeline` with model `sentence-transformers/all-mpnet-base-v2` the result is not normalized. Please refer to https://huggingface.co/sentence-transformers/all-mpnet-base-v2
What we get from pipeline is exactly like `sentence_embeddings` after `mean_pooling` and then need to run `F.normalize` - only after normalization the embeddings are equivalent to result of `encode()` from `SentenceTransformer` python package.
I searched myself for implementation of such normalization in Go packages and unfortunately didn't find any, the most closest stuff was in https://github.com/sugarme/gotch with their rich tensor functionality (but still `normalize` function is missing there). So I implemented it myself based on algorithm provided here:
https://pytorch.org/docs/stable/_modules/torch/nn/functional.html#normalize
The code is not long but required a lot of reverse engineering and debugging, so I will be glad to save time of other developers.
Please let me know if you want to add this functionality, either as new `normalize` function, or directly as output from `NewFeatureExtractionPipeline` (maybe with boolean param if to normalize).

**Comment by sedletsky-f5** [comment]
sounds good! do you want me to send you this function or how to proceed?

**Comment by RJKeevil** [comment]
Either works for me, can submit a draft pr or just post the code here?

**Comment by RJKeevil** [comment]
Hi @sedletsky-f5 can you push to a branch on your own github repo, and then start a pull request to our repo via the github interface?  I dont think github is set up for open pushes, and it doesnt seem to like me adding permissions to your specific user.

**Comment by sedletsky-f5** [comment]
@RJKeevil , I followed the regular procedure with github - cloned the repo, made a change and committed it but when I tried to push it, I got above error. it is a matter of permissions that you can give me in this repo via `Settings -> Access -> Collaborators and teams` (URL path `/settings/access`). Please give me `Role: Write`, it should be enough to push and then you can review and approve. Thanks

**Comment by RJKeevil** [comment]
Hi, i've invited you for write access. I think normally its better if you push to your own repo/fork ala https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request but for this its all good, just please make sure you push to a branch and not to main. Thanks again.

**Comment by sedletsky-f5** [comment]
thanks @RJKeevil , yes I am familiar with collaboration in github/gitlab :-)
created PR https://github.com/knights-analytics/hugot/pull/21 - please review

**Comment by sedletsky-f5** [comment]
I am just wondering how to explain all dependencies and installation details for the used `gotch` package. You can see how they did it in another package using it: https://github.com/sugarme/transformer/
Please suggest and I will document it.

**Comment by sedletsky-f5** [comment]
and another question is about pipeline - do you have in this repo and if so, how to trigger it to test no regression?

**Comment by RJKeevil** [comment]
yep we have a full suite, take a look at hugot_test and github actions  - run tests workflow. I'll get back to you tomorrow on the rest, 1am for me here :)

**Comment by sedletsky-f5** [comment]
I had issues with pipeline and it looked very complicated to update Dockerfile to always support this option, so I decided to solve it with conditional build (only those ones interested in normalization should use `-tags GOTCH` after all pre-requisites are installed like in my case). Running now pipeline again to see it is not impacted by this PR.

**Comment by RJKeevil** [comment]
Ok great, looking good. We can take it from here, we'll see if we can replicate just what we need from gotch in native go before merging.  That will help keep complexity down for users, and should avoid another round of the C tax.

**Comment by sedletsky-f5** [comment]
Thanks @RJKeevil , just let me know when you plan to merge so I will see if to use it or external normalization function in my current project.

**Comment by RJKeevil** [comment]
Will do, aiming for a merge this week

**Comment by riccardopinosio** [comment]
@sedletsky-f5 I have added a new commit for normalization that simplifies things somewhat, now it normalizes the onnxruntime output vectors in native go by implementing the formula at https://pytorch.org/docs/stable/generated/torch.nn.functional.normalize.html without need for gotch. I tested it against your previous gotch implementation and sentence transformers and it provides identical results.
It's less general than your gotch implementation of course but i think it makes sense to implement the special case here, and then eventually if we see that we want to move to tensor operations maybe rewrite it in gonum or gotch if we need a pytorch backend.

**Comment by sedletsky-f5** [comment]
It is really a great improvement, @riccardopinosio and I will try now with my example (to see the results are the same). Also I added 1 review comment for your commit. Thanks

**Comment by riccardopinosio** [comment]
@sedletsky-f5 I answered your comment :) let me know if it's clear. Also if this works for you, maybe you can squash the commits, so when I merge it's your contribution.

**Comment by sedletsky-f5** [comment]
@RJKeevil and @riccardopinosio can we advance with the merge (all comments are resolved) and tag? It would remove the complexity of `gotch` in my project.

**Comment by RJKeevil** [comment]
Perhaps if you can complete your testing of the 1.20 PR today, Riccardo and I can look at running the tests and releasing v0.1.1 tomorrow with both included?

**Comment by sedletsky-f5** [comment]
I found more places to fix for v1.20 and it might take some time. So maybe it would be better to not make dependency between these 2 efforts (and release already working normalization). While waiting for pipeline, I can make context switch to the second one :-) thanks

**Comment by riccardopinosio** [comment]
Merged to main
